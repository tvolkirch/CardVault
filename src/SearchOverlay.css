#search-overlay
{
    position: fixed;
    top: 0;
    display: none;
    height: 100vh;
    width: 100vw;
    background:rgba(255, 255, 255, 0.95);
    color: #333;
    margin: 0;
    padding: 1rem;
    z-index: 2;
}

#search-overlay .close-button
{
    font-size: 2rem;
    font-weight: bold;
    padding-left: 1rem;
    padding-right: 1rem;
    display: inline-block;
}

#search-overlay .close-icon:hover
{
    cursor: pointer;
}

#searchText
{
    line-height: 1.5rem;
    font-size: 1.25rem;
    border: 2px solid #333;
	-moz-border-radius: 0.25rem;
	-webkit-border-radius: 0.25rem;
	border-radius: 0.25rem;
	margin: 1rem 0;
}

#searchButton
{
    line-height: 1.5rem;
    font-size: 1.25rem;
    margin-left: 0.5rem;
}

.invisible
{
    background:rgba(255, 255, 255, 0) !important;
}
