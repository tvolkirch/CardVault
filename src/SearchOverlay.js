import React, { Component } from "react";

import CardService from "./api/CardService";
import Loading from "./Loading";
import SearchResults from "./SearchResults";

import { getPressedKeyStatus } from "./lib/EventManager";

import "./SearchOverlay.css";

class SearchOverlay extends Component
{
    constructor()
    {
        super();

        this.state = {
            loading: false,
            cards: [],
            count: -1,
            hasSearched: false
        };

        this.closeButton = React.createRef();
        this.searchText = React.createRef();
        this.searchButton = React.createRef();
    }

    componentDidMount()
    {
       /* some content initially doesn't exist - can
          only add event listener when element exists */

       if (this.closeButton.current)
       {
          this.closeButton.current.addEventListener("keydown", this.handleCloseButtonEvent);
       }

       if (this.searchText.current)
       {
          this.searchText.current.addEventListener("keydown", this.handleSearchTextEvent);
       }

       if (this.searchButton.current)
       {
          this.searchButton.current.addEventListener("keydown", this.handleSearchButtonEvent);
       }
    }

    componentDidUpdate()
    {
       var cardArray = document.getElementsByClassName("matched-card-name");

       if ( cardArray && cardArray.length > 0 )
       {
          for (let i = 0; i < cardArray.length; i++ )
          {
             cardArray[i].addEventListener("keydown", this.handleCardLinkEvent);
          }
       }
    }

    componentWillUnmount()
    {
       /* some content initially doesn't exist - can
          only add event listener when element exists */

       if (this.closeButton.current)
       {
          this.closeButton.current.removeEventListener("keydown", this.handleCloseButtonEvent);
       }

       if (this.searchText.current)
       {
          this.searchText.current.removeEventListener("keydown", this.handleSearchTextEvent);
       }

       if (this.searchButton.current)
       {
          this.searchButton.current.removeEventListener("keydown", this.handleSearchButtonEvent);
       }

       var cardArray = document.getElementsByClassName("matched-card-name");

       if ( cardArray && cardArray.length > 0 )
       {
          for (let i = 0; i < cardArray.length; i++ )
          {
             cardArray[i].removeEventListener("keydown", this.handleCardLinkEvent);
          }
       }
    }

    handleCloseButtonEvent = (eve) =>
    {
       const matchedCardList = document.getElementById("matchedCardList");

       this.handleCloseButton(eve, matchedCardList);
    }

    handleCloseButton(eve, matchedCardList)
    {
       const originatorElement = eve.target;
       const originatorId = originatorElement.getAttribute("id");
       const keyChecker = getPressedKeyStatus(eve);
       var lastTabbedControl = this.searchButton.current;
       var cardArray = [];

       if (matchedCardList)
       {
          cardArray = matchedCardList.getElementsByClassName("matched-card-name");
       }

       if ( cardArray && cardArray.length > 0 )
       {
          lastTabbedControl = cardArray[ cardArray.length - 1 ];
       }

       if ( originatorId === "closeSearchOverlay" )
       {
          if ( keyChecker.isEnter || keyChecker.isSpace )
          {
             this.closeSearchOverlay();
          }

          /* keep tabbing on search overlay */

          if ( keyChecker.isTab && eve.shiftKey )
          {
             eve.preventDefault();
             lastTabbedControl.focus();
          }
       }
    }

    handleSearchTextEvent = (eve) =>
    {
       const originatorElement = eve.target;
       const originatorId = originatorElement.getAttribute("id");

       const keyChecker = getPressedKeyStatus(eve);

       if ( originatorId === "searchText" && keyChecker.isEnter )
       {
          this.searchButton.current.click();
       }
    }

    handleSearchButtonEvent = (eve) =>
    {
       const matchedCardList = document.getElementById("matchedCardList");

       this.handleSearchButton(eve, matchedCardList);
    }

    handleSearchButton(eve, matchedCardList)
    {
       const originatorElement = eve.target;
       const originatorId = originatorElement.getAttribute("id");
       const keyChecker = getPressedKeyStatus(eve);
       var cardArray = [];

       if (matchedCardList)
       {
          cardArray = matchedCardList.getElementsByClassName("matched-card-name");
       }

       if ( originatorId === "searchButton"
           && (keyChecker.isEnter || keyChecker.isSpace) )
       {
          this.searchButton.current.click();
       }

       /* keep tabbing on search overlay */

       if ( !cardArray || cardArray.length === 0 )
       {
          if ( keyChecker.isTab && !eve.shiftKey )
          {
             eve.preventDefault();
             this.closeButton.current.focus();
          }
       }
    }

    handleCardLinkEvent = (eve) =>
    {
       const matchedCardList = document.getElementById("matchedCardList");

       this.handleCardLink(eve, matchedCardList);
    }

    handleCardLink(eve, matchedCardList)
    {
       const originatorElement = eve.target;
       const keyChecker = getPressedKeyStatus(eve);
       var cardArray = [];

       if (matchedCardList)
       {
          cardArray = matchedCardList.getElementsByClassName("matched-card-name");
       }

       if ( keyChecker.isEnter || keyChecker.isSpace )
       {
          originatorElement.click();
          return;
       }

       /* keep tabbing on search overlay */

       var lastTabbedControl = null;
       var lastTabbedControlIndex = "";
       const originatorIndex = originatorElement.getAttribute("data");

       if ( cardArray && cardArray.length > 0 )
       {
          lastTabbedControl = cardArray[ cardArray.length - 1 ];
          lastTabbedControlIndex = lastTabbedControl.getAttribute("data");
       }

       if ( keyChecker.isTab && !eve.shiftKey && originatorIndex === lastTabbedControlIndex )
       {
          eve.preventDefault();
          this.closeButton.current.focus();
       }
    }

    closeSearchOverlay()
    {
        var overlay = document.getElementById("search-overlay");
        var bodyTags = document.getElementsByTagName("body");
        var bodyTag = bodyTags[0];

        overlay.style.display = "none";
        bodyTag.style.overflow = "visible";  // show scrollbar again
    }

    search()
    {
        var searchElement = document.getElementById("searchText");

        this.searchApi(searchElement);
    }

    searchApi(searchElement)
    {
        this.setState( {loading: true} );

        const cardService = new CardService();
        var searchPromise = cardService.cardApi("find", searchElement.value);
        var count = 0;
        var cards = [];

        if (searchPromise && searchPromise.then)
        {
           searchPromise.then(response =>
           {
              count = response.data._totalCount;
              cards = response.data.cards;

              this.setState( {loading: false, cards: cards, count: count, hasSearched: true} );
           });
        }
        else
        {
           this.setState( {loading: false, cards: cards, count: count, hasSearched: true} );
        }
    }

    render()
    {
        var searchResults = "";
        var loader = "";

        if (this.state.loading)
        {
            loader = <Loading top={50} left={200} />;
        }

        if (this.state.hasSearched)
        {
           searchResults = <SearchResults
              count={this.state.count}
              cards={this.state.cards} />
        }

        return (
<div id="search-overlay" data-testid="search-overlay">
    {loader}
    <div id="closeSearchOverlay" className="close-button" ref={this.closeButton}
         data-testid="closeSearchOverlay"
         tabIndex="0" onClick={(eve) => this.closeSearchOverlay()}>
        <span className="close-icon" aria-label="close button" role="button">&#8855;</span>
    </div>
    <div id="search-form">
        <input type="search" size="25" name="searchText" id="searchText"
            data-testid="searchText"
            ref={this.searchText} placeholder="Search by Card Name"/>
        <input type="button" value="Search" id="searchButton"
            ref={this.searchButton} onClick={(eve) => this.search()} />
    </div>
    {searchResults}
</div>
        );
    }
}

export default SearchOverlay;