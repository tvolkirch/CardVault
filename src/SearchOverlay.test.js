import React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent } from "@testing-library/react";

import SearchOverlay from "./SearchOverlay";

jest.mock("./api/CardService");

it("test SearchOverlay rendering after mock search", () =>
{
   const searchOverlay = renderer.create(<SearchOverlay />);
   const self = searchOverlay.root.instance;
   var searchTextElem = document.createElement("div");

   searchTextElem.value = "mock";
   self.searchApi(searchTextElem);

   expect(searchOverlay.toJSON()).toMatchSnapshot();
});

it("test SearchOverlay rendering after failed search", () =>
{
   const searchOverlay = renderer.create(<SearchOverlay />);
   const root = searchOverlay.root;
   const self = root.instance;
   var searchTextElem = document.createElement("div");

   searchTextElem.value = "";
   self.searchApi(searchTextElem);

   const message = root.findByProps({id: "search-results"}).children;

   expect(message).toEqual(["No card names were found that matched the search text."]);
});

it("test handleCloseButton for pressed Enter key", () =>
{
   const searchOverlay = new SearchOverlay();
   var closeButtonElem = document.createElement("div");

   closeButtonElem.setAttribute("id", "closeSearchOverlay");

   var mockEve =
   {
      code: "Enter",
      target: closeButtonElem
   };

   searchOverlay.closeSearchOverlay = jest.fn( () => { return; } );
   searchOverlay.handleCloseButton(mockEve, null);

   expect(searchOverlay.closeSearchOverlay).toHaveBeenCalled();
});

it("test handleCloseButton for pressed space bar", () =>
{
   const searchOverlay = new SearchOverlay();
   var closeButtonElem = document.createElement("div");

   closeButtonElem.setAttribute("id", "closeSearchOverlay");

   var mockEve =
   {
      code: "Space",
      target: closeButtonElem
   };

   searchOverlay.closeSearchOverlay = jest.fn( () => { return; } );
   searchOverlay.handleCloseButton(mockEve, null);

   expect(searchOverlay.closeSearchOverlay).toHaveBeenCalled();
});

it("test handleCloseButton for unsupported key", () =>
{
   const searchOverlay = new SearchOverlay();
   var closeButtonElem = document.createElement("div");

   closeButtonElem.setAttribute("id", "closeSearchOverlay");

   var mockEve =
   {
      code: "KeyE",
      target: closeButtonElem
   };

   searchOverlay.closeSearchOverlay = jest.fn( () => { return; } );
   searchOverlay.handleCloseButton(mockEve, null);

   expect(searchOverlay.closeSearchOverlay).not.toHaveBeenCalled();
});

it("test handleCloseButton for shift/tab with no matching cards", () =>
{
   const searchOverlay = new SearchOverlay();
   var closeButtonElem = document.createElement("div");
   var tabControlElem = document.createElement("div");

   closeButtonElem.setAttribute("id", "closeSearchOverlay");
   tabControlElem.focus = jest.fn( () => { return; } );

   var mockEve =
   {
      code: "Tab",
      shiftKey: true,
      target: closeButtonElem,
      preventDefault: () => { return; }
   };

   searchOverlay.searchButton = {};
   searchOverlay.searchButton.current = tabControlElem;
   searchOverlay.handleCloseButton(mockEve, null);

   expect(tabControlElem.focus).toHaveBeenCalled();
});


it("test handleCloseButton for tab with no matching cards", () =>
{
   const searchOverlay = new SearchOverlay();
   var closeButtonElem = document.createElement("div");
   var tabControlElem = document.createElement("div");

   closeButtonElem.setAttribute("id", "closeSearchOverlay");
   tabControlElem.focus = jest.fn( () => { return; } );

   var mockEve =
   {
      code: "Tab",
      shiftKey: false,
      target: closeButtonElem,
      preventDefault: () => { return; }
   };

   searchOverlay.searchButton = {};
   searchOverlay.searchButton.current = tabControlElem;
   searchOverlay.handleCloseButton(mockEve, null);

   expect(tabControlElem.focus).not.toHaveBeenCalled();
});

it("test handleCloseButton for shift/tab with a matching card", () =>
{
   const searchOverlay = new SearchOverlay();
   var closeButtonElem = document.createElement("div");
   var tabControlElem = document.createElement("div");
   var card = document.createElement("div");
   var matchedCardListElem = document.createElement("div");

   closeButtonElem.setAttribute("id", "closeSearchOverlay");
   tabControlElem.focus = jest.fn( () => { return; } );
   card.setAttribute("class", "matched-card-name");
   card.focus = jest.fn( () => { return; } );
   matchedCardListElem.appendChild(card);

   var mockEve =
   {
      code: "Tab",
      shiftKey: true,
      target: closeButtonElem,
      preventDefault: () => { return; }
   };

   searchOverlay.searchButton = {};
   searchOverlay.searchButton.current = tabControlElem;
   searchOverlay.handleCloseButton(mockEve, matchedCardListElem);

   expect(searchOverlay.searchButton.current.focus).not.toHaveBeenCalled();
   expect(card.focus).toHaveBeenCalled();
});

it("test handleCloseButton for tab with a matching card", () =>
{
   const searchOverlay = new SearchOverlay();
   var closeButtonElem = document.createElement("div");
   var tabControlElem = document.createElement("div");
   var card = document.createElement("div");
   var matchedCardListElem = document.createElement("div");

   closeButtonElem.setAttribute("id", "closeSearchOverlay");
   card.setAttribute("class", "matched-card-name");
   card.focus = jest.fn( () => { return; } );
   matchedCardListElem.appendChild(card);

   var mockEve =
   {
      code: "Tab",
      shiftKey: false,
      target: closeButtonElem,
      preventDefault: () => { return; }
   };

   searchOverlay.searchButton = {};
   searchOverlay.searchButton.current = document.createElement("div");
   searchOverlay.handleCloseButton(mockEve, matchedCardListElem);

   expect(card.focus).not.toHaveBeenCalled();
});

it("test handleSearchText for pressed Enter key", () =>
{
   const searchOverlay = new SearchOverlay();
   var searchTextElem = document.createElement("div");
   var searchButtonElem = document.createElement("div");

   searchTextElem.setAttribute("id", "searchText");
   searchButtonElem.click = jest.fn( () => { return; } );

   var mockEve =
   {
      code: "Enter",
      target: searchTextElem
   };

   searchOverlay.searchButton = {};
   searchOverlay.searchButton.current = searchButtonElem;
   searchOverlay.handleSearchTextEvent(mockEve);

   expect(searchOverlay.searchButton.current.click).toHaveBeenCalled();
});

it("test handleSearchText for unsupported key", () =>
{
   const searchOverlay = new SearchOverlay();
   var searchTextElem = document.createElement("div");
   var searchButtonElem = document.createElement("div");

   searchTextElem.setAttribute("id", "searchText");
   searchButtonElem.click = jest.fn( () => { return; } );

   var mockEve =
   {
      code: "Space",
      target: searchTextElem
   };

   searchOverlay.searchButton = {};
   searchOverlay.searchButton.current = searchButtonElem;
   searchOverlay.handleSearchTextEvent(mockEve);

   expect(searchOverlay.searchButton.current.click).not.toHaveBeenCalled();
});

it("test handleSearchButton for pressed Enter key", () =>
{
   const searchOverlay = new SearchOverlay();
   var searchButtonElem = document.createElement("div");

   searchButtonElem.setAttribute("id", "searchButton");
   searchButtonElem.click = jest.fn( () => { return; } );

   var mockEve =
   {
      code: "Enter",
      target: searchButtonElem
   };

   searchOverlay.searchButton = {};
   searchOverlay.searchButton.current = searchButtonElem;
   searchOverlay.handleSearchButton(mockEve, null);

   expect(searchOverlay.searchButton.current.click).toHaveBeenCalled();
});

it("test handleSearchButton for pressed Space key", () =>
{
   const searchOverlay = new SearchOverlay();
   var searchButtonElem = document.createElement("div");

   searchButtonElem.setAttribute("id", "searchButton");
   searchButtonElem.click = jest.fn( () => { return; } );

   var mockEve =
   {
      code: "Space",
      target: searchButtonElem
   };

   searchOverlay.searchButton = {};
   searchOverlay.searchButton.current = searchButtonElem;
   searchOverlay.handleSearchButton(mockEve, null);

   expect(searchOverlay.searchButton.current.click).toHaveBeenCalled();
});

it("test handleSearchButton for tab", () =>
{
   const searchOverlay = new SearchOverlay();
   var searchButtonElem = document.createElement("div");
   var closeSearchOverlayElem = document.createElement("div");

   searchButtonElem.setAttribute("id", "searchButton");
   searchButtonElem.click = jest.fn( () => { return; } );
   closeSearchOverlayElem.focus = jest.fn( () => { return; } );

   var mockEve =
   {
      code: "Tab",
      shiftKey: false,
      target: searchButtonElem,
      preventDefault: () => { return; }
   };

   searchOverlay.closeButton = {};
   searchOverlay.closeButton.current = closeSearchOverlayElem;
   searchOverlay.searchButton = {};
   searchOverlay.searchButton.current = searchButtonElem;
   searchOverlay.handleSearchButton(mockEve, null);

   expect(searchOverlay.searchButton.current.click).not.toHaveBeenCalled();
   expect(searchOverlay.closeButton.current.focus).toHaveBeenCalled();
});

it("test handleCardLink for pressed Enter key", () =>
{
   const searchOverlay = new SearchOverlay();
   var cardNameLinkElem = document.createElement("div");

   cardNameLinkElem.click = jest.fn( () => { return; } );

   var mockEve =
   {
      code: "Enter",
      target: cardNameLinkElem
   };

   searchOverlay.handleCardLink(mockEve, null);

   expect(cardNameLinkElem.click).toHaveBeenCalled();
});

it("test handleCardLink for pressed space bar", () =>
{
   const searchOverlay = new SearchOverlay();
   var cardNameLinkElem = document.createElement("div");

   cardNameLinkElem.click = jest.fn( () => { return; } );

   var mockEve =
   {
      code: "Space",
      target: cardNameLinkElem
   };

   searchOverlay.handleCardLink(mockEve, null);

   expect(cardNameLinkElem.click).toHaveBeenCalled();
});

it("test handleCardLink for tab from last card link", () =>
{
   const searchOverlay = new SearchOverlay();
   var cardNameLinkElem = document.createElement("div");
   var closeSearchOverlayElem = document.createElement("div");
   var matchedCardListElem = document.createElement("div");

   cardNameLinkElem.setAttribute("class", "matched-card-name");
   cardNameLinkElem.setAttribute("data", "0");
   cardNameLinkElem.click = jest.fn( () => { return; } );
   matchedCardListElem.appendChild(cardNameLinkElem);
   closeSearchOverlayElem.focus = jest.fn( () => { return; } );

   var mockEve =
   {
      code: "Tab",
      shiftKey: false,
      target: cardNameLinkElem,
      preventDefault: () => { return; }
   };

   searchOverlay.closeButton = {};
   searchOverlay.closeButton.current = closeSearchOverlayElem;
   searchOverlay.handleCardLink(mockEve, matchedCardListElem);

   expect(cardNameLinkElem.click).not.toHaveBeenCalled();
   expect(searchOverlay.closeButton.current.focus).toHaveBeenCalled();
});

it("test handleCardLink for tab from card link that isn't last", () =>
{
   const searchOverlay = new SearchOverlay();
   var cardNameLinkElem1 = document.createElement("div");
   var cardNameLinkElem2 = document.createElement("div");
   var closeSearchOverlayElem = document.createElement("div");
   var matchedCardListElem = document.createElement("div");

   cardNameLinkElem1.setAttribute("data", "0");
   cardNameLinkElem2.setAttribute("data", "1");
   cardNameLinkElem1.click = jest.fn( () => { return; } );
   matchedCardListElem.appendChild(cardNameLinkElem1);
   matchedCardListElem.appendChild(cardNameLinkElem2);
   closeSearchOverlayElem.focus = jest.fn( () => { return; } );

   var mockEve =
   {
      code: "Tab",
      shiftKey: false,
      target: cardNameLinkElem1,
      preventDefault: () => { return; }
   };

   searchOverlay.closeButton = {};
   searchOverlay.closeButton.current = closeSearchOverlayElem;
   searchOverlay.handleCardLink(mockEve, matchedCardListElem);

   expect(cardNameLinkElem1.click).not.toHaveBeenCalled();
   expect(closeSearchOverlayElem.focus).not.toHaveBeenCalled();
});

it("test to close card overlay", () =>
{
   const { getByTestId } = render(<SearchOverlay />);
   var closeButtonElem = getByTestId("closeSearchOverlay");

   fireEvent.click(closeButtonElem);

   var searchOverlayElem = getByTestId("search-overlay");

   expect(searchOverlayElem.style.display).toBe("none");
});
