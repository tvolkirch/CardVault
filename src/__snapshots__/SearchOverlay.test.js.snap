// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SearchOverlay rendering after mock search 1`] = `
<div
  data-testid="search-overlay"
  id="search-overlay"
>
  
  <div
    className="close-button"
    data-testid="closeSearchOverlay"
    id="closeSearchOverlay"
    onClick={[Function]}
    tabIndex="0"
  >
    <span
      aria-label="close button"
      className="close-icon"
      role="button"
    >
      ⊗
    </span>
  </div>
  <div
    id="search-form"
  >
    <input
      data-testid="searchText"
      id="searchText"
      name="searchText"
      placeholder="Search by Card Name"
      size="25"
      type="search"
    />
    <input
      id="searchButton"
      onClick={[Function]}
      type="button"
      value="Search"
    />
  </div>
  <div>
    <div
      data-testid="search-results"
      id="search-results"
      tabIndex="0"
    >
      The search text matched 1 card name.  Click on the card name to display the card.
    </div>
    <ul
      data-testid="matchedCardList"
      id="matchedCardList"
    >
      <li>
        <div
          className="matched-card-name"
          data={0}
          data-testid="matchedCardName0"
          id="matchedCardName0"
          onClick={[Function]}
          tabIndex="0"
        >
          Mocked Card Search
        </div>
        <div
          className="card-overlay"
          data-testid="cardOverlay0"
          id="cardOverlay0"
        >
          <div
            className="close-button"
            data={0}
            data-testid="closeCardOverlay0"
            id="closeCardOverlay0"
            onClick={[Function]}
            tabIndex="0"
          >
            <span
              aria-label="close button"
              className="close-icon"
              role="button"
            >
              ⊗
            </span>
          </div>
          <div
            className="flex-card-wrapper"
          >
            <div
              className="card-content"
              tabIndex="0"
            >
              <img
                alt="Elder Scrolls Card: Mocked Card Search"
                className="card-image"
                src="#"
              />
            </div>
            <div
              className="card-content card-text"
              data={0}
              tabIndex="0"
            >
              <div
                className="card-name"
              >
                Mocked Card Search
              </div>
              <div>
                <div
                  className="flex-card-row"
                >
                  <span
                    className="left-column"
                  >
                    Type:
                  </span>
                  <span
                    className="right-column"
                  >
                    Mocked Search
                  </span>
                </div>
                <div
                  className="flex-card-row"
                >
                  <span
                    className="left-column"
                  >
                    Set:
                  </span>
                  <span
                    className="right-column"
                  >
                    Mocked Search Set
                  </span>
                </div>
              </div>
              <div
                className="card-rules"
              >
                This is mocked data for unit testing of card search.
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
